### PR TITLE
Address TF/ DDPG low return

### LINKED ISSUE #1077: Too less avg_retrun
Formula for off_policy_method:

total_timeseteps= n_epochs * n_epoch_cycles * batch_size 

then if 
n_epochs=1400
n_epoch_cycles=20
batch_size=64
min_buffer_size=10^6
then total_timesteps=1400 * 20 * 64=1,792,000

I obtained the graph as shown in figure for DDPG_Walker2d-v3. It is showing very less average return. But most of the research papers shows ~2500 average_return on 1 million timesteps. How to set the parameters to get near about results.
My Code:

import gym
import tensorflow as tf
import time
from garage.experiment import run_experiment
from garage.np.exploration_strategies import OUStrategy
from garage.replay_buffer import SimpleReplayBuffer
from garage.tf.algos import DDPG
from garage.tf.envs import TfEnv
from garage.tf.experiment import LocalTFRunner
from garage.tf.policies import ContinuousMLPPolicy
from garage.tf.q_functions import ContinuousMLPQFunction
import random
from datetime import datetime, timedelta
import numpy as np
import os

def run_task(snapshot_config, *_):
    """Run task."""

    with LocalTFRunner(snapshot_config=snapshot_config) as runner:
        env=gym.make('Walker2d-v3')
        env = TfEnv(env)
        action_noise = OUStrategy(env.spec, sigma=0.2)

        policy = ContinuousMLPPolicy(env_spec=env.spec,
                                     hidden_sizes=[400, 300],
                                     hidden_nonlinearity=tf.nn.relu,
                                     output_nonlinearity=tf.nn.tanh)

        qf = ContinuousMLPQFunction(env_spec=env.spec,
                                    hidden_sizes=[400,300],
                                    hidden_nonlinearity=tf.nn.relu)

        replay_buffer = SimpleReplayBuffer(env_spec=env.spec,
                                           size_in_transitions=int(1e6),
                                           time_horizon=100)

        ddpg = DDPG(env_spec=env.spec,
                    policy=policy,
                    policy_lr=1e-4,
                    qf_lr=1e-3,
                    qf=qf,
                    replay_buffer=replay_buffer,
                    target_update_tau=1e-3,
                    n_train_steps=50,
                    discount=0.99,
                    buffer_batch_size=64,
                    n_epoch_cycles=20,
                    min_buffer_size=int(1e6),
                    exploration_strategy=action_noise,
                    policy_optimizer=tf.train.AdamOptimizer,
                    qf_weight_decay=0.01,
                    qf_optimizer=tf.train.AdamOptimizer)

        runner.setup(algo=ddpg, env=env)

        runner.train(n_epochs=2000, n_epoch_cycles=20, batch_size=64)

sed=[21]
for difsed in range(1):
    i=0
    seed=sed[difsed]
    #for i in range(2):
    start_time=time.time()
    run_experiment(
        run_task,
        snapshot_mode='last',
        seed=seed,
        exp_name=str(seed)+"_"+str(i),
        log_dir=r"/home/surabhi/Downloads/github/garage/result/ddpg/walk-v22/"+str(seed)+"/"+str(i)+"/"
    )
        #print("Time: ",timedelta(seconds=time.time()-start_time))
    file=open(r"/home/surabhi/Downloads/github/garage/result/ddpg/walk-v22/time.txt","a")
    file.write('seed '+str(seed)+' itr '+str(i)+' start '+str(start_time)+' elapsed '+str(timedelta(seconds=time.time()-start_time))+"\n")
    file.close()

f

I found that it starts evaluation from 782epoch ( bcz 1000000//(20*64) ). Hence https://github.com/rlworkgroup/garage/blob/master/src/garage/tf/algos/ddpg.py condition on line 272 will be true from this epoch and policy optimization will be started from this point onward. Is this the reason for not getting results?

Another question i want to ask is: why this evaluation loop (line 271 of DDPG.py) is repeated n_train_steps (training steps) times? what is the purpose of evaluating n_train_steps times. Is this doing a kind of rollout repeated n_train_steps times? Where length of rollout is either end of episode or a trajectory of length =batch_size=64. (Line 173 of https://github.com/rlworkgroup/garage/blob/master/src/garage/sampler/off_policy_vectorized_sampler.py#L66 ?

### PR DESCRIPTION
**Pull request by irisliucy**
This PR addresses #1077 from with the following investigation directions. 

Investigation of low performance of TF/ DDPG: 
- [x]  Add noise to the actions (no noise added to testing).

- [x] Add warm-up period `start_steps=1000`(Select action randomly before the first 1000 time steps, afterwards select action based on policy; update the policy after the warm-up period).

- [ ] Use gaussian noise instead of OU noise (The original paper proposed OU noise but in recent papers such as TD3/  D4PG, there is no performance benefit for OU noise).

### CODE REVIEW

File: src/garage/np/policies/uniform_random_policy.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        Args:
+            observations(list): Observations from the environment.
+
+        Returns:
+            np.ndarray: Actions with noise.
+            List[dict]: Arbitrary policy state information (agent_info).
+
+        """
+        return [self._env_spec.action_space.sample()
+                for obs in observations], dict()
```
Comment (ryanjulian) [inline src/garage/np/policies/uniform_random_policy.py:62]:
```suggestion
        return [self._env_spec.action_space.sample()
                for _ in range(len(observations))], dict()
```

File: src/garage/tf/algos/ddpg.py [hunk 2]
```diff
@@ -55,6 +51,8 @@ class DDPG(RLAlgorithm):
             clip_return].
         max_action (float): Maximum action magnitude.
         reward_scale (float): Reward scale.
+        exploration_policy_sigma (float): Action noise sigma.
```
Comment (ryanjulian) [inline src/garage/tf/algos/ddpg.py:54]: please construct the exploration strategy in the launcher

File: src/garage/tf/algos/ddpg.py [hunk 3]
```diff
@@ -276,34 +279,49 @@ def train(self, runner):
...
             for cycle in range(self._steps_per_epoch):
-                runner.step_path = runner.obtain_episodes(runner.step_itr)
-                self.train_once(runner.step_itr, runner.step_path)
+                # Get action randomly from environment within warm-up steps.
+                # Afterwards, get action from policy.
+                if runner.step_itr >= self._start_steps:
+                    runner.step_path = runner.obtain_episodes(
+                        runner.step_itr, agent_update=self.exploration_policy)
+                else:
+                    uniform_random_policy = UniformRandomPolicy(self.env_spec)
```
Comment (ryanjulian) [inline src/garage/tf/algos/ddpg.py:293]: this should be constructed in the launcher file

File: src/garage/tf/algos/ddpg.py [hunk 4]
```diff
@@ -276,34 +279,49 @@ def train(self, runner):
...
+                        num_eps=self._num_evaluation_episodes)
                     last_returns = log_performance(runner.step_itr,
                                                    eval_episodes,
                                                    discount=self._discount)
                 runner.step_itr += 1
 
         return np.mean(last_returns)
 
-    def train_once(self, itr, episodes):
+    def train_once(self, itr):
```
Comment (ryanjulian) [inline src/garage/tf/algos/ddpg.py:318]:
```suggestion
    def _train_once(self, itr):
```

**Review by ryanjulian — COMMENTED** [review]
LGTM. What are the results like?

**Review by krzentner — COMMENTED** [review]
This looks pretty good, but please address Ryan's comments.

**Comment by irisliucy** [comment]
This is the result after adding noise to action and warm-up period:

**Comment by ryanjulian** [comment]
can you please compare to openai baselines (or tf_agents or spinningup) for us?
